File: day25.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

const checkLen = 20
var expected [checkLen]int

const (
	cpy = iota
	inc
	dec
	jnz
	out
)

type Arg interface {
	Value() int
	Name() string
}

type LiteralArg int

func (a LiteralArg) Value() int {
	return int(a)
}

func (a LiteralArg) Name() string {
	panic(fmt.Sprintf("Misused literal: %v", int(a)))
}

var regs map[string]int

type RegisterArg string

func (a RegisterArg) Value() int {
	return regs[string(a)]
}

func (a RegisterArg) Name() string {
	return string(a)
}

func newArg(s string) Arg {
	x, err := strconv.Atoi(s)
	if err == nil {
		return LiteralArg(x)
	}
	return RegisterArg(s)
}

type Cmd struct {
	op  int
	x, y Arg
}

func main() {
	for i := 0; i < checkLen; i++ {
		expected[i] = i % 2
	}

	cmds := []Cmd{}

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		var x, y string

		n, _ := fmt.Sscanf(line, "cpy %s %s", &x, &y)
		if n == 2 {
			cmds = append(cmds, Cmd{cpy, newArg(x), newArg(y)})
			continue
		}

		n, _ = fmt.Sscanf(line, "inc %s", &x)
		if n == 1 {
			cmds = append(cmds, Cmd{inc, RegisterArg(x), LiteralArg(0)})
			continue
		}

		n, _ = fmt.Sscanf(line, "dec %s", &x)
		if n == 1 {
			cmds = append(cmds, Cmd{dec, RegisterArg(x), LiteralArg(0)})
			continue
		}

		n, _ = fmt.Sscanf(line, "jnz %s %s", &x, &y)
		if n == 2 {
			cmds = append(cmds, Cmd{jnz, newArg(x), newArg(y)})
			continue
		}

		n, _ = fmt.Sscanf(line, "out %s", &x)
		if n == 1 {
			cmds = append(cmds, Cmd{out, newArg(x), LiteralArg(0)})
			continue
		}

		log.Fatal("Unknown command: %s", line)
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	for a0 := 1; a0 < 1e6; a0++ {
		output := run(cmds, a0)
		//fmt.Println(a0, output)
		if output == expected {
			fmt.Println(a0)
			break
		}
	}
}

func run(cmds []Cmd, a0 int) [checkLen]int {
	regs = make(map[string]int)
	regs["a"] = a0

	output := make([]int, 0, checkLen)
	ip := 0
	for 0 <= ip && ip <= len(cmds) {
		cmd := cmds[ip]

		jumped := false
		switch cmd.op {
		case out:
			output = append(output, cmd.x.Value())
			if len(output) == checkLen {
				a := [checkLen]int{}
				copy(a[:checkLen], output)
				return a
			}
		case cpy:
			regs[cmd.y.Name()] = cmd.x.Value()
		case jnz:
			if cmd.x.Value() != 0 {
				ip += cmd.y.Value()
				jumped = true
			}
		case inc:
			regs[cmd.x.Name()]++
		case dec:
			regs[cmd.x.Name()]--
		default:
			log.Fatal("Unknown cmd: %v", cmd)
		}

		if !jumped {
			ip++
		}
	}

	rv := [checkLen]int{}
	return rv
}

File: day25.py
import collections
import fileinput
import itertools
import re
import sys

regs = None

cmds = []
for line in fileinput.input():
    line = line.strip()

    m = re.search(r"cpy ([-\w]+) ([-\w]+)", line)
    if m:
        cmds.append(['cpy', m.group(1), m.group(2)])
        continue

    m = re.search(r"inc (\w+)", line)
    if m:
        cmds.append(['inc', m.group(1)])
        continue

    m = re.search(r"dec (\w+)", line)
    if m:
        cmds.append(['dec', m.group(1)])
        continue

    m = re.search(r"jnz ([-\w]+) ([-\w]+)", line)
    if m:
        cmds.append(['jnz', m.group(1), m.group(2)])
        continue

    m = re.search(r"out ([-\w]+)", line)
    if m:
       cmds.append(['out', m.group(1)])
       continue

    assert not line

def value(x):
    try:
        return int(x)
    except ValueError:
        return regs[x]

def cpy(x, y):
    if re.match("[a-z]", y):
        regs[y] = value(x)
    return None

def jnz(x, y):
    if value(x) != 0:
        return value(y)
    return None

def inc(x):
    if re.match("[a-z]", x):
        regs[x] += 1
    return None

def dec(x):
    if re.match("[a-z]", x):
        regs[x] -= 1
    return None

def run(a0):
   global regs
   regs = collections.defaultdict(int)
   regs["a"] = a0

   output = []
   ip = 0
   while 0 <= ip < len(cmds):
       cmd = cmds[ip]

       if cmd[0] == "out":
          output.append(value(cmd[1]))
          if len(output) >= 20:
             return output
       else:
          f = globals()[cmd[0]]
          r = f(*cmd[1:])
          if r is None:
              ip += 1
          else:
              ip += r

       if ip < 0 or ip >= len(cmds):
           break

for a0 in itertools.count(1):
   sys.stderr.write(".")
   s20 = list(itertools.islice(run(a0), 20))
   #print s20
   #print [0, 1] * 10
   #print
   if s20 == [0, 1] * 10:
      print a0
      break


